rust/src/shards/ws.rs

```diff
@@ -0,0 +1,323 @@
+  }
+
+  fn setParam(&mut self, index: i32, value: &Var) -> Result<(), &str> {
+    self.client.set_param(index, value);
+    Ok(())
+  }
+
+  fn getParam(&mut self, index: i32) -> Var {
+    self.client.get_param(index)
+  }
+
+  fn warmup(&mut self, context: &Context) -> Result<(), &str> {
+    self.client.warmup(context)
+  }
+
+  fn cleanup(&mut self) -> Result<(), &str> {
+    self.client.cleanup()
+  }
+
+  fn activate(&mut self, context: &Context, input: &Var) -> Result<Var, &str> {
+    Ok(run_blocking(self, context, input))
+  }
+}
+
+impl BlockingShard for ReadString {
+  fn activate_blocking(&mut self, _context: &Context, _input: &Var) -> Result<Var, &str> {
+    let ws = self.client.activate()?;
+
+    loop {
+      let msg = ws.read_message().map_err(|e| {
+        shlog!("{}", e);
+        "Failed to read message."
+      })?;
+
+      match msg {
+        Message::Text(text) => {
+          self.text = text.into();
+          return Ok(self.text.0);
+        }
+        Message::Binary(_) => {
+          return Err("Received binary message, expected text.");
+        }
+        Message::Close(_) => {
+          return Err("Received close message.");
+        }
+        Message::Ping(_) => {
+          // shlog_debug!("Received ping message, sending pong.");
+          ws.write_message(Message::Pong(vec![])).map_err(|e| {
+            shlog!("{}", e);
+            "Failed to write message."
+          })?;
+        }
+        Message::Pong(_) => {
+          return Err("Received pong message, expected text.");
+        }
+        _ => return Err("Invalid message type."),
+      }
+    }
+  }
+}
+
+#[derive(Default)]
+struct WriteString {
+  client: ClientUser,
+}
+
+impl Shard for WriteString {
+  fn registerName() -> &'static str
+  where
+    Self: Sized,
+  {
+    cstr!("WS.WriteString")
+  }
+
+  fn hash() -> u32
+  where
+    Self: Sized,
+  {
+    compile_time_crc32::crc32!("WS.WriteString-rust-0x20200101")
+  }
+
+  fn name(&mut self) -> &str {
+    "WS.WriteString"
+  }
+
+  fn inputTypes(&mut self) -> &Types {
+    &STRING_TYPES
+  }
+
+  fn outputTypes(&mut self) -> &Types {
+    &STRING_TYPES
+  }
+
+  fn parameters(&mut self) -> Option<&Parameters> {
+    Some(&WS_CLIENT_USER_PARAMS)
+  }
+
+  fn requiredVariables(&mut self) -> Option<&ExposedTypes> {
+    self.client.required_variables()
+  }
+
+  fn setParam(&mut self, index: i32, value: &Var) -> Result<(), &str> {
+    self.client.set_param(index, value);
+    Ok(())
+  }
+
+  fn getParam(&mut self, index: i32) -> Var {
+    self.client.get_param(index)
+  }
+
+  fn warmup(&mut self, context: &Context) -> Result<(), &str> {
+    self.client.warmup(context)
+  }
+
+  fn cleanup(&mut self) -> Result<(), &str> {
+    self.client.cleanup()
+  }
+
+  fn activate(&mut self, _context: &Context, input: &Var) -> Result<Var, &str> {
+    // TODO make this run on a worker thread but not using BlockingShard as would overlap/crash
+    let ws = self.client.activate()?;
```

Comment:
Where is the overlap, is it in the websocket library itself?

Reply:
no it was here, both were using await